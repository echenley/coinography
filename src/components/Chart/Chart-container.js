// @flow

import { connect, type Connector } from 'react-redux'
import { fetchChartData } from '../../actions'
import {
  getCandleChartData,
  getChartDataLoadingState,
  getChartType,
  getCurrency,
  getGranularity,
  getLineChartData,
} from '../../selectors'
import Chart, { type Props } from './Chart'
import type { State } from '../../types'

function mapStateToProps(state: State) {
  const chartType = getChartType(state)

  return {
    chartData:
      chartType === 'line'
        ? getLineChartData(state)
        : getCandleChartData(state),
    chartType,
    currency: getCurrency(state),
    granularity: getGranularity(state),
    loadingState: getChartDataLoadingState(state),
  }
}

const connector: Connector<{}, Props> = connect(mapStateToProps, {
  fetchChartData,
})

export default connector(Chart)
