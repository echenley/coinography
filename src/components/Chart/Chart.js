// @flow

import * as React from 'react'
import moment from 'moment'
import type { AxiosPromise } from 'axios'
import { Dimensions, StyleSheet, Text, View } from 'react-native'
import {
  VictoryAxis,
  VictoryCandlestick,
  VictoryChart,
  VictoryLine,
  VictoryTheme,
} from 'victory-native'
import { Icon, Spinner } from 'native-base'
import type {
  CandleData,
  ChartType,
  Currency,
  Granularity,
  HistoricData,
  LineData,
  LoadingState,
} from '../../types'

export type Props = {
  chartData: LineData | CandleData | null,
  chartType: ChartType,
  currency: Currency,
  fetchChartData: (
    currency: Currency,
    granularity: Granularity,
    silent?: boolean
  ) => AxiosPromise<HistoricData>,
  granularity: Granularity,
  loadingState: LoadingState,
}

type State = {
  width: number,
}

class Chart extends React.Component<Props, State> {
  state = {
    width: Dimensions.get('window').width - 30,
  }

  intervalId: ?IntervalID

  componentDidMount() {
    Dimensions.addEventListener('change', this.updateWidth)

    this.fetchData()
    this.resetPolling()
  }

  componentDidUpdate(prevProps: Props) {
    const { currency, granularity } = this.props

    if (
      prevProps.currency !== currency ||
      prevProps.granularity !== granularity
    ) {
      this.fetchData()
      this.resetPolling()
    }
  }

  componentWillUnmount() {
    Dimensions.removeEventListener('change', this.updateWidth)
  }

  resetPolling = () => {
    if (this.intervalId) {
      clearInterval(this.intervalId)
    }

    this.intervalId = setInterval(() => this.fetchData(true), 10000)
  }

  updateWidth = (e: Object) => {
    this.setState({ width: e.window.width - 30 })
  }

  fetchData = (silent?: boolean = false) => {
    const { currency, granularity } = this.props
    this.props.fetchChartData(currency, granularity, silent)
  }

  render() {
    const { chartType, chartData, loadingState } = this.props
    const currentYear = new Date().getFullYear()

    let chart = null
    let overlay = null

    if (loadingState !== 'success') {
      let icon = null
      if (loadingState === 'loading') {
        icon = <Spinner color="white" />
      } else if (loadingState === 'failure') {
        icon = (
          <View style={styles.failed}>
            <Icon color="white" name="warning" style={styles.failedIcon} />
            <Text style={styles.failedText}>Failed to fetch data</Text>
          </View>
        )
      }

      overlay = <View style={styles.loadingOverlay}>{icon}</View>
    }

    if (!chartData) {
      return <View style={styles.empty}>{overlay}</View>
    }

    if (chartType === 'candle') {
      chart = (
        <VictoryCandlestick
          candleColors={{ positive: 'green', negative: 'red' }}
          data={chartData}
        />
      )
    } else {
      chart = (
        <VictoryLine data={chartData} style={{ data: { stroke: 'blue' } }} />
      )
    }

    return (
      <View>
        <VictoryChart
          domainPadding={{ x: 5, y: 25 }}
          padding={{ left: 65, right: 10, bottom: 40, top: 0 }}
          scale={{ x: 'time' }}
          theme={VictoryTheme.material}
          width={this.state.width}
        >
          <VictoryAxis
            fixLabelOverlap
            tickFormat={(date, i, dates) => {
              const adjacentIndex = i === 0 ? 1 : i - 1

              // ticks are >= 24 hours apart
              if (Math.abs(dates[adjacentIndex] - date) > 1000 * 60 * 60 * 24) {
                const year = date.getFullYear()
                const format = year !== currentYear ? 'l' : 'MMM D'
                return moment(date).format(format)
              }

              const showMinutes = date.getMinutes() !== 0
              return moment(date).format(showMinutes ? 'LT' : 'h A')
            }}
          />
          <VictoryAxis
            dependentAxis
            tickFormat={value => {
              const decimals = value < 1000 ? 2 : 0

              return value.toLocaleString('en-US', {
                style: 'currency',
                currency: 'USD',
                minimumFractionDigits: decimals,
                maximumFractionDigits: decimals,
              })
            }}
          />
          {chart}
        </VictoryChart>
        {overlay}
      </View>
    )
  }
}

const styles = StyleSheet.create({
  empty: {
    height: 400,
  },
  loadingOverlay: {
    position: 'absolute',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'rgba(0, 0, 0, 0.2)',
    width: '100%',
    height: '100%',
  },
  failed: {
    display: 'flex',
    alignItems: 'center',
  },
  failedIcon: {
    color: 'white',
    fontSize: 80,
  },
  failedText: {
    color: 'white',
    fontSize: 18,
    fontWeight: 'bold',
  },
})

export default Chart
