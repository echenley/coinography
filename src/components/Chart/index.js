// @flow

import Chart from './Chart-container'
export default Chart
