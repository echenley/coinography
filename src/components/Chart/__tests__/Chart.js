import * as React from 'react'
import { Dimensions } from 'react-native'
import { shallow } from 'enzyme'
import Chart from '../Chart'

jest.useFakeTimers()

Dimensions.addEventListener = jest.fn()
Dimensions.removeEventListener = jest.fn()

const candleData = [
  {
    x: '2018-03-03T22:05:00.000Z',
    open: 11307,
    close: 11309.64,
    high: 11315.01,
    low: 11301.04,
  },
  {
    x: '2018-03-03T22:00:00.000Z',
    open: 11298.98,
    close: 11307,
    high: 11307,
    low: 11290.05,
  },
  {
    x: '2018-03-03T21:55:00.000Z',
    open: 11290.25,
    close: 11298.98,
    high: 11298.98,
    low: 11270,
  },
  {
    x: '2018-03-03T21:50:00.000Z',
    open: 11284.35,
    close: 11290.24,
    high: 11291.05,
    low: 11280,
  },
  {
    x: '2018-03-03T21:45:00.000Z',
    open: 11293.5,
    close: 11284.35,
    high: 11293.5,
    low: 11280.54,
  },
]

const lineData = [
  { x: '2018-03-03T22:05:00.000Z', y: 11309.64 },
  { x: '2018-03-03T22:00:00.000Z', y: 11307 },
  { x: '2018-03-03T21:55:00.000Z', y: 11298.98 },
  { x: '2018-03-03T21:50:00.000Z', y: 11290.24 },
  { x: '2018-03-03T21:45:00.000Z', y: 11284.35 },
]

function renderComponent(props) {
  return shallow(
    <Chart
      chartData={null}
      chartType="candle"
      currency="BTC-USD"
      fetchChartData={() => Promise.resolve([])}
      granularity={300}
      loadingState="loading"
      {...props}
    />
  )
}

describe('<Chart />', () => {
  afterEach(() => {
    Dimensions.addEventListener.mockClear()
    Dimensions.removeEventListener.mockClear()
  })

  it('renders correctly while loading', () => {
    expect(renderComponent().getElement()).toMatchSnapshot()
  })

  it('renders correctly while loading with stale data', () => {
    expect(
      renderComponent({
        chartData: candleData,
      }).getElement()
    ).toMatchSnapshot()
  })

  it('renders correctly in failure state', () => {
    expect(
      renderComponent({
        loadingState: 'failure',
      }).getElement()
    ).toMatchSnapshot()
  })

  it('renders a candle chart correctly', () => {
    expect(
      renderComponent({
        chartData: candleData,
        loadingState: 'success',
      }).getElement()
    ).toMatchSnapshot()
  })

  it('renders a line chart correctly', () => {
    expect(
      renderComponent({
        chartData: lineData,
        loadingState: 'success',
      }).getElement()
    ).toMatchSnapshot()
  })

  it('adds and removes Dimensions listener on mount and unmount', () => {
    const component = renderComponent()
    const instance = component.instance()

    expect(instance.intervalId).toBeDefined()
    expect(Dimensions.addEventListener).toHaveBeenCalledWith(
      'change',
      instance.updateWidth
    )

    component.unmount()

    expect(Dimensions.removeEventListener).toHaveBeenCalledWith(
      'change',
      instance.updateWidth
    )
  })

  it('updates state.width on orientation change', () => {
    const component = renderComponent()
    const { updateWidth } = component.instance()

    updateWidth({
      window: {
        width: 800,
      },
    })

    component.update()

    expect(component.state('width')).toBe(770)
  })

  it('fetches data on mount and on update', () => {
    const fetchChartData = jest.fn()
    const component = renderComponent({
      fetchChartData,
    })

    expect(fetchChartData).toHaveBeenCalledWith('BTC-USD', 300, false)

    component.setProps({
      currency: 'ETH-USD',
    })

    expect(fetchChartData).toHaveBeenCalledWith('ETH-USD', 300, false)
  })

  it('polls for chart data every 10 seconds', () => {
    const fetchChartData = jest.fn()
    renderComponent({
      fetchChartData,
    })


    jest.advanceTimersByTime(20000)

    expect(fetchChartData).toHaveBeenCalledTimes(3)
    expect(fetchChartData).toHaveBeenCalledWith('BTC-USD', 300, false)
    expect(fetchChartData).toHaveBeenCalledWith('BTC-USD', 300, true)
  })
})
