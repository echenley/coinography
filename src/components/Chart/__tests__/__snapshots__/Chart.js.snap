// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chart /> renders a candle chart correctly 1`] = `
<View>
  <VictoryChart
    domainPadding={
      Object {
        "x": 5,
        "y": 25,
      }
    }
    padding={
      Object {
        "bottom": 40,
        "left": 65,
        "right": 10,
        "top": 0,
      }
    }
    scale={
      Object {
        "x": "time",
      }
    }
    theme="VictoryTheme.material"
    width={720}
  >
    <VictoryAxis
      fixLabelOverlap={true}
      tickFormat={[Function]}
    />
    <VictoryAxis
      dependentAxis={true}
      tickFormat={[Function]}
    />
    <VictoryCandlestick
      candleColors={
        Object {
          "negative": "red",
          "positive": "green",
        }
      }
      data={
        Array [
          Object {
            "close": 11309.64,
            "high": 11315.01,
            "low": 11301.04,
            "open": 11307,
            "x": "2018-03-03T22:05:00.000Z",
          },
          Object {
            "close": 11307,
            "high": 11307,
            "low": 11290.05,
            "open": 11298.98,
            "x": "2018-03-03T22:00:00.000Z",
          },
          Object {
            "close": 11298.98,
            "high": 11298.98,
            "low": 11270,
            "open": 11290.25,
            "x": "2018-03-03T21:55:00.000Z",
          },
          Object {
            "close": 11290.24,
            "high": 11291.05,
            "low": 11280,
            "open": 11284.35,
            "x": "2018-03-03T21:50:00.000Z",
          },
          Object {
            "close": 11284.35,
            "high": 11293.5,
            "low": 11280.54,
            "open": 11293.5,
            "x": "2018-03-03T21:45:00.000Z",
          },
        ]
      }
    />
  </VictoryChart>
</View>
`;

exports[`<Chart /> renders a line chart correctly 1`] = `
<View>
  <VictoryChart
    domainPadding={
      Object {
        "x": 5,
        "y": 25,
      }
    }
    padding={
      Object {
        "bottom": 40,
        "left": 65,
        "right": 10,
        "top": 0,
      }
    }
    scale={
      Object {
        "x": "time",
      }
    }
    theme="VictoryTheme.material"
    width={720}
  >
    <VictoryAxis
      fixLabelOverlap={true}
      tickFormat={[Function]}
    />
    <VictoryAxis
      dependentAxis={true}
      tickFormat={[Function]}
    />
    <VictoryCandlestick
      candleColors={
        Object {
          "negative": "red",
          "positive": "green",
        }
      }
      data={
        Array [
          Object {
            "x": "2018-03-03T22:05:00.000Z",
            "y": 11309.64,
          },
          Object {
            "x": "2018-03-03T22:00:00.000Z",
            "y": 11307,
          },
          Object {
            "x": "2018-03-03T21:55:00.000Z",
            "y": 11298.98,
          },
          Object {
            "x": "2018-03-03T21:50:00.000Z",
            "y": 11290.24,
          },
          Object {
            "x": "2018-03-03T21:45:00.000Z",
            "y": 11284.35,
          },
        ]
      }
    />
  </VictoryChart>
</View>
`;

exports[`<Chart /> renders correctly in failure state 1`] = `
<View
  style={
    Object {
      "height": 400,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.2)",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
        }
      }
    >
      <Styled(Icon)
        color="white"
        name="warning"
        style={
          Object {
            "color": "white",
            "fontSize": 80,
          }
        }
        virtual={undefined}
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        Failed to fetch data
      </Text>
    </View>
  </View>
</View>
`;

exports[`<Chart /> renders correctly while loading 1`] = `
<View
  style={
    Object {
      "height": 400,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.2)",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Styled(Spinner)
      color="white"
      virtual={undefined}
    />
  </View>
</View>
`;

exports[`<Chart /> renders correctly while loading with stale data 1`] = `
<View>
  <VictoryChart
    domainPadding={
      Object {
        "x": 5,
        "y": 25,
      }
    }
    padding={
      Object {
        "bottom": 40,
        "left": 65,
        "right": 10,
        "top": 0,
      }
    }
    scale={
      Object {
        "x": "time",
      }
    }
    theme="VictoryTheme.material"
    width={720}
  >
    <VictoryAxis
      fixLabelOverlap={true}
      tickFormat={[Function]}
    />
    <VictoryAxis
      dependentAxis={true}
      tickFormat={[Function]}
    />
    <VictoryCandlestick
      candleColors={
        Object {
          "negative": "red",
          "positive": "green",
        }
      }
      data={
        Array [
          Object {
            "close": 11309.64,
            "high": 11315.01,
            "low": 11301.04,
            "open": 11307,
            "x": "2018-03-03T22:05:00.000Z",
          },
          Object {
            "close": 11307,
            "high": 11307,
            "low": 11290.05,
            "open": 11298.98,
            "x": "2018-03-03T22:00:00.000Z",
          },
          Object {
            "close": 11298.98,
            "high": 11298.98,
            "low": 11270,
            "open": 11290.25,
            "x": "2018-03-03T21:55:00.000Z",
          },
          Object {
            "close": 11290.24,
            "high": 11291.05,
            "low": 11280,
            "open": 11284.35,
            "x": "2018-03-03T21:50:00.000Z",
          },
          Object {
            "close": 11284.35,
            "high": 11293.5,
            "low": 11280.54,
            "open": 11293.5,
            "x": "2018-03-03T21:45:00.000Z",
          },
        ]
      }
    />
  </VictoryChart>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.2)",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <Styled(Spinner)
      color="white"
      virtual={undefined}
    />
  </View>
</View>
`;
